### ISSUE TITLE
[API Proposal]:  ContainerConfiguration WithAssemblies Retains the Actual Collection; and Can Become Cleared

### ISSUE THREAD

**Issue by steevcoco** [dotnet/runtime#101126]
### Background and motivation

When invoking a `ContainerConfiguration` with `WithAssemblies(IEnumerable<Assembly>)`; and passing a collection instance; then the `Collection` itself is retained. If the collection becomes modified, then the `ContainerConfiguration` changes.

This may be intended, or correct; yet the method takes an `IEnumerable`, and it was surprising to discover that no copy was made.

### API Proposal

The documentation for `ContainerConfiguration.WithAssemblies` states:

Adds part types from the specified collection of assemblies to the container.

It infers that "part types" are added now; and not really the collection itself.

Either the documentation or the implementation could be changed.

### API Usage

```csharp
[Export]
public sealed class MyExport
{
    public static void Assert(bool isFailure)
    {
        ConventionBuilder conventions = new();
        conventions.ForType(typeof(MyExport)).Export();
        ContainerConfiguration configuration = new();
        _ = configuration.WithDefaultConventions(conventions);
        
        HashSet<Assembly> assemblies = [ typeof(MyExport).Assembly ];
        _ = configuration.WithAssemblies(assemblies);
        if (isFailure)
            assemblies.Clear();
        
        CompositionHost host = configuration.CreateContainer();
        MyExport[] exports = host.GetExports<MyExport>().ToArray();
        if (isFailure)
            Debug.Assert(0 == exports.Length);
        else
            Debug.Assert(1 == exports.Length);
    }


    public static void Main()
    {
        Assert(false);
        Assert(true);
    }
}

```

**Comment by buyaa-n** [comment]
@steevcoco the issue doesn't look like an API suggestion, so removing the `api-suggestion` label. 

I am not expert in this area so not sure I follow your point. Are you saying that once the `assemblies` collection added to the configuration even if the `assemblies` content is cleared the configuration should keep the parts added with the collection?

> Either the documentation or the implementation could be changed.

This library is old, legacy assembly for which we would not accept changes unless necessary. Most likely we will not fix the implementation as that could break others. Doc could be changed if needed.

**Comment by steevcoco** [comment]
You know, it's not exactly as odd as I described; and given the state of that Mef, it should just be closed. … Most api methods that take an enumerable seem to iterate it right away, and since the method documentation doesn't explicitly say so, then I only found out by luck. 

It's the configuration, and it's not the container; so that behavior seems acceptable. In a perfect world, the documentation would actually go on to say it's not enumerated until composition happens .. 

Thanks for your time! 😁
